@masterleinad testing this PR (and kokkos-kernels develop branch) with Trilinos (Cuda build, like reproducer in #5189 ) I get this error in kokkos-kernels:

```
[ 22%] Building CXX object kokkos-kernels/src/CMakeFiles/kokkoskernels.dir/impl/generated_specializations_cpp/trsm/Blas3_trsm_eti_DOUBLE_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDAUVMSPACE.cpp.o
cd /ascldap/users/ndellin/trilinos/Trilinos/Build/Weaver-PR5198/kokkos-kernels/src && /home/projects/ppc64le-pwr9-nvidia/openmpi/4.0.1/gcc/7.2.0/cuda/10.1.105/bin/mpicxx  -I/ascldap/users/ndellin/trilinos/Trilinos/Build/Weaver-PR5198 -I/ascldap/users/ndellin/trilinos/Trilinos/Build/Weaver-PR5198/kokkos-kernels/src -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/impl -I/ascldap/users/ndellin/trilinos/Trilinos/Build/Weaver-PR5198/kokkos-kernels/src/impl -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/impl/tpls -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/blas -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/blas/impl -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/sparse -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/sparse/impl -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/graph -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/graph/impl -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/batched -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/batched/dense -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/batched/dense/impl -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/batched/sparse -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/batched/sparse/impl -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/common -I/ascldap/users/ndellin/trilinos/Trilinos/Build/Weaver-PR5198/kokkos/algorithms/src -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos/algorithms/src -I/ascldap/users/ndellin/trilinos/Trilinos/Build/Weaver-PR5198/kokkos/containers/src -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos/containers/src -I/ascldap/users/ndellin/trilinos/Trilinos/Build/Weaver-PR5198/kokkos/core/src -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos/core/src -I/ascldap/users/ndellin/trilinos/Trilinos/Build/Weaver-PR5198/kokkos -I/ascldap/users/ndellin/trilinos/Trilinos/kokkos/core/src/../../tpls/desul/include -I/ascldap/users/projects/ppc64le-pwr9-nvidia/cuda/10.1.105/include -I/ascldap/users/projects/ppc64le-pwr9/metis/5.0.1/gcc/7.2.0/include -pedantic -Wall -mcpu=power9 -mtune=power9  -expt-extended-lambda -arch=sm_70  -O3 -DNDEBUG -fPIC -std=c++14 -o CMakeFiles/kokkoskernels.dir/impl/generated_specializations_cpp/trsm/Blas3_trsm_eti_DOUBLE_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDAUVMSPACE.cpp.o -c /ascldap/users/ndellin/trilinos/Trilinos/Build/Weaver-PR5198/kokkos-kernels/src/impl/generated_specializations_cpp/trsm/Blas3_trsm_eti_DOUBLE_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDAUVMSPACE.cpp

/ascldap/users/ndellin/trilinos/Trilinos/kokkos/core/src/Kokkos_CopyViews.hpp(3762): error: no suitable user-defined conversion from "const Kokkos::View<const double **, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::CudaUVMSpace::execution_space, Kokkos::CudaUVMSpace::memory_space>, Kokkos::MemoryTraits<1U>>" to "std::enable_if_t<true, Kokkos::View<double **, std::conditional_t<true, Kokkos::Cuda::array_layout, Kokkos::Cuda::array_layout>, Kokkos::Device<Kokkos::HostSpace::execution_space, Kokkos::Cuda::memory_space>, std::conditional_t<false, void, Kokkos::Experimental::DefaultViewHooks>>>" exists
          detected during:
            instantiation of "std::enable_if_t<std::is_same<Kokkos::View<T, P...>::memory_space, Kokkos::View<T, P...>::HostMirror::memory_space>::value, Kokkos::View<T, P...>::HostMirror> Kokkos::create_mirror_view(const Kokkos::View<T, P...> &) [with T=const double **, P=<Kokkos::LayoutLeft, Kokkos::Device<Kokkos::CudaUVMSpace::execution_space, Kokkos::CudaUVMSpace::memory_space>, Kokkos::MemoryTraits<1U>>]"
/ascldap/users/ndellin/trilinos/Trilinos/kokkos-kernels/src/blas/impl/KokkosBlas3_trsm_spec.hpp(133): here
            instantiation of "void KokkosBlas::Impl::TRSM<AViewType, BViewType, false, true>::trsm(const char *, const char *, const char *, const char *, BViewType::const_value_type &, const AViewType &, const BViewType &) [with AViewType=Kokkos::View<const double **, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::CudaUVMSpace::execution_space, Kokkos::CudaUVMSpace::memory_space>, Kokkos::MemoryTraits<1U>>, BViewType=Kokkos::View<double **, Kokkos::LayoutLeft, Kokkos::Device<Kokkos::CudaUVMSpace::execution_space, Kokkos::CudaUVMSpace::memory_space>, Kokkos::MemoryTraits<1U>>]"
/ascldap/users/ndellin/trilinos/Trilinos/Build/Weaver-PR5198/kokkos-kernels/src/impl/generated_specializations_cpp/trsm/Blas3_trsm_eti_DOUBLE_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDAUVMSPACE.cpp(52): here

1 error detected in the compilation of "/tmp/tmpxft_00024581_00000000-6_Blas3_trsm_eti_DOUBLE_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDAUVMSPACE.cpp1.ii".
make[2]: *** [kokkos-kernels/src/CMakeFiles/kokkoskernels.dir/impl/generated_specializations_cpp/trsm/Blas3_trsm_eti_DOUBLE_LAYOUTLEFT_EXECSPACE_CUDA_MEMSPACE_CUDAUVMSPACE.cpp.o] Error 1
```

--- a/containers/src/Kokkos_DynRankView.hpp
+++ b/containers/src/Kokkos_DynRankView.hpp
@@ -2131,9 +2131,11 @@ create_mirror_view(
 // Create a mirror view in host space
 template <class T, class... P>
 inline std::enable_if_t<
-    std::is_same<
-        typename DynRankView<T, P...>::memory_space,
-        typename DynRankView<T, P...>::HostMirror::memory_space>::value,
+    (std::is_same<
+         typename DynRankView<T, P...>::memory_space,
+         typename DynRankView<T, P...>::HostMirror::memory_space>::value &&
+     std::is_same<typename DynRankView<T, P...>::data_type,
+                  typename DynRankView<T, P...>::HostMirror::data_type>::value),
     typename DynRankView<T, P...>::HostMirror>
 create_mirror_view(const Kokkos::DynRankView<T, P...>& src) {
   return src;
@@ -2141,9 +2143,12 @@ create_mirror_view(const Kokkos::DynRankView<T, P...>& src) {
 
 template <class T, class... P>
 inline std::enable_if_t<
-    !std::is_same<
-        typename DynRankView<T, P...>::memory_space,
-        typename DynRankView<T, P...>::HostMirror::memory_space>::value,
+    !(std::is_same<
+          typename DynRankView<T, P...>::memory_space,
+          typename DynRankView<T, P...>::HostMirror::memory_space>::value &&
+      std::is_same<
+          typename DynRankView<T, P...>::data_type,
+          typename DynRankView<T, P...>::HostMirror::data_type>::value),
     typename DynRankView<T, P...>::HostMirror>
 create_mirror_view(const Kokkos::DynRankView<T, P...>& src) {
   return Kokkos::create_mirror(src);

--- a/core/src/Kokkos_CopyViews.hpp
+++ b/core/src/Kokkos_CopyViews.hpp
@@ -3754,19 +3754,27 @@ create_mirror_view(const Space&, const Kokkos::View<T, P...>& src,
 }  // namespace Impl
 
 template <class T, class... P>
-std::enable_if_t<std::is_same<typename Kokkos::View<T, P...>::memory_space,
-                              typename Kokkos::View<
-                                  T, P...>::HostMirror::memory_space>::value,
-                 typename Kokkos::View<T, P...>::HostMirror>
+std::enable_if_t<
+    std::is_same<
+        typename Kokkos::View<T, P...>::memory_space,
+        typename Kokkos::View<T, P...>::HostMirror::memory_space>::value &&
+        std::is_same<
+            typename Kokkos::View<T, P...>::data_type,
+            typename Kokkos::View<T, P...>::HostMirror::data_type>::value,
+    typename Kokkos::View<T, P...>::HostMirror>
 create_mirror_view(const Kokkos::View<T, P...>& src) {
   return src;
 }
 
 template <class T, class... P>
-std::enable_if_t<!std::is_same<typename Kokkos::View<T, P...>::memory_space,
-                               typename Kokkos::View<
-                                   T, P...>::HostMirror::memory_space>::value,
-                 typename Kokkos::View<T, P...>::HostMirror>
+std::enable_if_t<
+    !(std::is_same<
+          typename Kokkos::View<T, P...>::memory_space,
+          typename Kokkos::View<T, P...>::HostMirror::memory_space>::value &&
+      std::is_same<
+          typename Kokkos::View<T, P...>::data_type,
+          typename Kokkos::View<T, P...>::HostMirror::data_type>::value),
+    typename Kokkos::View<T, P...>::HostMirror>
 create_mirror_view(const Kokkos::View<T, P...>& src) {
   return Kokkos::create_mirror(src);
 }

